File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        total = [root.val]
        def maxPath(root):
            if not root:
                return 0  

            left = max(maxPath(root.left), 0)
            right = max(maxPath(root.right), 0)
            
            total[0] = max(total[0], root.val + left + right)
            return root.val + max(left, right)

        maxPath(root)
        return total[0]

File: Data Structures & Algorithms/depth-of-binary-tree/submission-0.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {number}
     */
    maxDepth(root) {
        if (!root) {
            return 0
        }
        let leftCount = 1;
        let rightCount = 1;
        if (root.left) {
            leftCount += this.maxDepth(root.left)
        }
        if (root.right) {
            rightCount += this.maxDepth(root.right)
        }

        return Math.max(leftCount, rightCount)
    }
}
